src/Castle.Core.Tests/Interceptors/AsyncInterceptor.cs

```diff
@@ -0,0 +1,41 @@
+﻿// Copyright 2004-2010 Castle Project - http://www.castleproject.org/
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace CastleTests.Interceptors
+{
+	using System.Threading.Tasks;
+	using Castle.DynamicProxy;
+
+	public class AsyncInterceptor : IInterceptor
+	{
+		public void Intercept(IInvocation invocation)
+		{
+			invocation.ReturnValue = InterceptAsyncMethod(invocation);
+		}
+
+		private static async Task InterceptAsyncMethod(IInvocation invocation)
+		{
+			// It all falls down when executing async before calling Proceed().
+			await Task.Delay(10).ConfigureAwait(false);
```

Comment:
What happens if you change the line above to:

```csharp
Task.Delay(10).GetAwaiter().GetResult();
```

Reply:
Managed to get test passing here: https://github.com/fir3pho3nixx/Core/pull/28. Maybe @StephenCleary can validate my approach as I borrowed some code from Nito.AsyncEx to make this work.